### PR TITLE
Feat/sonic wall

### PR DESCRIPTION
**Pull request by Adamowoc**

### CODE REVIEW

File: SonicWall/sonicwall-fw/ingest/parser.yml [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+          2: 'correlation'
+          3: 'action'
+        mapping:
+          parsed_event.message.type: cef.event_type
+        fallback: 'base event'
+
+  set_cef_extension_fields_cef:
+    actions:
+      - set:
+          cef: '{{parsed_event.message}}'
```
Comment (squioc) [inline SonicWall/sonicwall-fw/ingest/parser.yml:298]: As only `cef.type` is declared in your taxonomy, you can remove this part, as the result will be always empty.
```suggestion
```

File: SonicWall/sonicwall-fw/tests/test_app_control_detection_alert.json [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    },
+    "sonicwall": {
+      "fw": {
+        "appcat": "PROTOCOLS",
+        "cat": "0",
+        "catid": "77",
+        "cnt": "7800123",
+        "cs3Label": "WAN",
+        "cs4Label": "WORKSTATION",
+        "fw_action": "NA",
```
Comment (squioc) [inline SonicWall/sonicwall-fw/tests/test_app_control_detection_alert.json:71]: `NA` values should be discarded from the result. Please update the parser :pray:

File: SonicWall/sonicwall-fw/tests/test_app_control_detection_alert.json [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    "host": {
+      "network": {
+        "ingress": {
+          "bytes": 1240
+        }
+      }
+    },
+    "cef": {},
+    "network": {
+      "transport": "tcp/https"
```
Comment (squioc) [inline SonicWall/sonicwall-fw/tests/test_app_control_detection_alert.json:57]: [`network.transport`](https://www.elastic.co/guide/en/ecs/8.4/ecs-network.html#field-network-transport) should represent the transport layer (here `tcp`).
You should split the value into the `network.transport` and the [`network.protocol`](https://www.elastic.co/guide/en/ecs/8.4/ecs-network.html#field-network-protocol) to obtain:
```json
"network": {
   "transport": "tcp",
   "protocol": "https"
}
```
Be careful: others examples exposed that the second part can be a number. However, `network.protocol` should only hold a string.

File: SonicWall/sonicwall-fw/ingest/parser.yml [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+          destination.ip: '{{parsed_event.message.dst or parsed_event.message.dstV6}}'
+          destination.mac: '{{parsed_event.message.dmac}}'
+          destination.nat.ip: '{{parsed_event.message.cs2Label or parsed_event.message.natDstV6}}'
+          destination.nat.port: '{{parsed_event.message.dnpt}}'
+          destination.packets: '{{parsed_event.message.cn1Label}}'
+          destination.port: '{{parsed_event.message.dpt}}'
+          network.transport: '{{parsed_event.message.proto}}'
+          observer.egress.interface.name: '{{parsed_event.message.deviceOutboundInterface}}'
+          observer.ingress.interface.name: '{{parsed_event.message.deviceInboundInterface}}'
+          observer.type: 'firewall'
```
Comment (squioc) [inline SonicWall/sonicwall-fw/ingest/parser.yml:314]: In conflict with line 63

File: SonicWall/sonicwall-fw/ingest/parser.yml [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+          # process.id: '{{parsed_event.message.app}}'
+          process.entity_id: '{{parsed_event.message.appid}}'
+          process.name: '{{parsed_event.message.appName}}'
+          rule.id: '{{parsed_event.message.cs1}}'
```
Comment (squioc) [inline SonicWall/sonicwall-fw/ingest/parser.yml:318]: According to the tests, [`rule.name`](https://www.elastic.co/guide/en/ecs/8.4/ecs-rule.html#field-rule-name) seems more suitable.
```suggestion
          rule.name: '{{parsed_event.message.cs1}}'
```

File: SonicWall/sonicwall-fw/ingest/parser.yml [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+          cef: '{{parsed_event.message}}'
+
+  set_ecs_fields:
+    actions:
+      - set:
+          destination.bytes: '{{parsed_event.message.in}}'
+          destination.domain: '{{parsed_event.message.request}}'
+          destination.ip: '{{parsed_event.message.dst or parsed_event.message.dstV6}}'
+          destination.mac: '{{parsed_event.message.dmac}}'
+          destination.nat.ip: '{{parsed_event.message.cs2Label or parsed_event.message.natDstV6}}'
```
Comment (squioc) [inline SonicWall/sonicwall-fw/ingest/parser.yml:307]: Override line 107. Maybe you should merge the two lines.

File: SonicWall/sonicwall-fw/ingest/parser.yml [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+      - set:
+          cef: '{{parsed_event.message}}'
+
+  set_ecs_fields:
+    actions:
+      - set:
+          destination.bytes: '{{parsed_event.message.in}}'
+          destination.domain: '{{parsed_event.message.request}}'
+          destination.ip: '{{parsed_event.message.dst or parsed_event.message.dstV6}}'
+          destination.mac: '{{parsed_event.message.dmac}}'
```
Comment (squioc) [inline SonicWall/sonicwall-fw/ingest/parser.yml:306]: Duplicated with line 112

File: SonicWall/sonicwall-fw/ingest/parser.yml [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+          observer.type: 'firewall'
+          # process.id: '{{parsed_event.message.app}}'
+          process.entity_id: '{{parsed_event.message.appid}}'
+          process.name: '{{parsed_event.message.appName}}'
+          rule.id: '{{parsed_event.message.cs1}}'
+          source.bytes: '{{parsed_event.message.out}}'
+          source.ip: '{{parsed_event.message.src or parsed_event.message.srcV6}}'
+          source.mac: '{{parsed_event.message.smac}}'
+          source.nat.ip: '{{parsed_event.message.cs1Label or parsed_event.message.natSrcV6}}'
+          source.nat.port: '{{parsed_event.message.snpt}}'
```
Comment (squioc) [inline SonicWall/sonicwall-fw/ingest/parser.yml:323]: Override line 130 and 131.

File: SonicWall/sonicwall-fw/ingest/parser.yml [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+          observer.egress.interface.name: '{{parsed_event.message.deviceOutboundInterface}}'
+          observer.ingress.interface.name: '{{parsed_event.message.deviceInboundInterface}}'
+          observer.type: 'firewall'
+          # process.id: '{{parsed_event.message.app}}'
+          process.entity_id: '{{parsed_event.message.appid}}'
+          process.name: '{{parsed_event.message.appName}}'
+          rule.id: '{{parsed_event.message.cs1}}'
+          source.bytes: '{{parsed_event.message.out}}'
+          source.ip: '{{parsed_event.message.src or parsed_event.message.srcV6}}'
+          source.mac: '{{parsed_event.message.smac}}'
```
Comment (squioc) [inline SonicWall/sonicwall-fw/ingest/parser.yml:321]: Duplicated with line 128

File: SonicWall/sonicwall-fw/ingest/parser.yml [hunk 10]
```diff
@@ -0,0 +1,355 @@
...
+          source.nat.ip: '{{parsed_event.message.cs1Label or parsed_event.message.natSrcV6}}'
+          source.nat.port: '{{parsed_event.message.snpt}}'
+          source.packets: '{{parsed_event.message.cn2Label}}'
+          source.port: '{{parsed_event.message.spt}}'
```
Comment (squioc) [inline SonicWall/sonicwall-fw/ingest/parser.yml:325]: Duplicated with line 129

**Review by squioc — CHANGES_REQUESTED** [review]
Thanks for your work.

I spotted several issues in the parser. Please fix them.
I think you can merge custom stages `set_cef_extension_fields` and `set_ecs_fields`; there is several duplicated or overridden fields.

File: SonicWall/sonicwall-fw/_meta/smart-descriptions.json [hunk 11]
```diff
@@ -0,0 +1,10 @@
+[
+    {
+        "value": "{event.reason}",
+        "conditions": [
+            {
+                "field": "event.reason"
+            }
+        ]
+    }
```
Comment (squioc) [inline SonicWall/sonicwall-fw/_meta/smart-descriptions.json:9]: Could you add some others smart-descriptions?

File: SonicWall/sonicwall-fw/ingest/parser.yml [hunk 12]
```diff
@@ -0,0 +1,355 @@
...
+          # network.protocol: '{{parsed_event.message.app}}' is id
+          event.reason: '{{parsed_event.message.msg}}'
+          file.name: '{{parsed_event.message.fname}}'
+          file.size: '{{parsed_event.message.fsize}}'
+          source.ip: '{{parsed_event.message.src or parsed_event.message.sourceAddress}}'
+          source.user.name: '{{parsed_event.message.suser}}'
+          destination.domain: "{{parsed_event.message.get('destination DnsDomain')}}"
+          service.name: "{{parsed_event.message.get('destination ServiceName')}}"
+          destination.nat.ip: "{{parsed_event.message.get('destination Translated Address')}}"
+          destination.nat.port: "{{parsed_event.message.get('destination TranslatedPort')}}"
```
Comment (squioc) [inline SonicWall/sonicwall-fw/ingest/parser.yml:87]: Could you add and fill [`event.kind`](https://www.elastic.co/guide/en/ecs/8.4/ecs-allowed-values-event-kind.html), [`event.category`](https://www.elastic.co/guide/en/ecs/8.4/ecs-allowed-values-event-category.html) and [`event.type`](https://www.elastic.co/guide/en/ecs/8.4/ecs-allowed-values-event-type.html)?

File: SonicWall/README.md [hunk 13]
```diff
@@ -0,0 +1,7 @@
+# SonicWall
+
+## Description
+Sonic Wall is a cybersecurity company specialized in content control and network security.
+
+## Intakes
+*
```
Comment (PierrickV) [inline SonicWall/README.md:7]:
```suggestion
```

This part is empty, you might want to remove it

File: SonicWall/sonicwall-fw/_meta/fields.yml [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+    name: sonicwall.fw.cs5
+    description: "URL: Applicable only when Network Packet Capture System (NPCS Solera) is enabled, displays URL of an NPCS object"
+    type: keyword
+
+sonicwall.fw.cs5Label:
+    name: sonicwall.fw.cs5Label
+    description: "Pre-defined string indicating session type: Applies to Syslogs with an associated user session being tracked by the UTM. "
+    type: keyword
+
+sonicwall.fw.cs6:
```
Comment (PierrickV) [inline SonicWall/sonicwall-fw/_meta/fields.yml:67]: Those fields are CEF specific and might not contain the same information for each event. Please use `.get('<FIELD NAME>)` to collect the content of the field.

Comment (Adamowoc) [reply SonicWall/sonicwall-fw/_meta/fields.yml:67]: SonicWall does not conform to the CEF norm, you can see in their [documentation](https://www.sonicwall.com/techdocs/pdf/sonicos-6-5-4-log-events-reference-guide.pdf) that their fields do not correspond to the labels.

File: SonicWall/sonicwall-fw/ingest/parser.yml [hunk 15]
```diff
@@ -0,0 +1,1655 @@
...
+          destination.user.name: "{{parsed_event.message.duser}}"
+          event.action: '{{parsed_event.message.act or parsed_event.message.reason}}'
+          event.category: ['network']
+          event.end: "{{parse_end_date.date}}"
+          event.kind: 'event'
+          event.outcome: "{{parsed_event.message.outcome}}"
+          event.reason: '{{parsed_event.message.msg}}'
+          event.start:  "{{parse_start_date.date}}"
+          event.timezone: "{{parsed_event.message.dtz}}"
+          event.type : 'protocol'
```
Comment (squioc) [inline SonicWall/sonicwall-fw/ingest/parser.yml:103]: Must be an array
```suggestion
          event.type : ['protocol']
```

File: SonicWall/sonicwall-fw/ingest/parser.yml [hunk 16]
```diff
@@ -0,0 +1,1655 @@
...
+          source.domain: "{{parsed_event.message.get('sourceDns Domain')}}"
+          source.ip: '{{parsed_event.message.src or parsed_event.message.sourceAddress or parsed_event.message.srcV6}}'
+          source.mac: '{{parsed_event.message.smac}}'
+          source.nat.ip: "{{parsed_event.message.cs1Label or parsed_event.message.natSrcV6 or parsed_event.message.get('source Translated Address')}}"
+          source.nat.port: "{{parsed_event.message.snpt or parsed_event.message.get('source TranslatedPort')}}"
+          source.packets: '{{parsed_event.message.cn2Label}}'
+          source.port: '{{parsed_event.message.spt}}'
+          source.user.name: '{{parsed_event.message.suser}}'
+          user_agent.original: "{{parsed_event.message.get('requestClient Application')}}"
+          user.name: '{{parsed_event.message.susr}}'
```
Comment (squioc) [inline SonicWall/sonicwall-fw/ingest/parser.yml:168]: According to the tests, this field may hold a username with a user domain. I suggest splitting the value with a grok stage like [this example](https://github.com/SEKOIA-IO/intake-formats/blob/59859d1b7048a0a24f00b7bb82940b8dc9118fee/Azure/azure-windows/ingest/parser.yml#L20-L30)

File: SonicWall/sonicwall-fw/tests/test_file_transfer_result.json [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+  "expected": {
+    "message": "CEF:0|SonicWall|NSa 4700|7.0.1-5000-R3000|1460|Capture ATP File Transfer Result|5|cat=1 gcat=3 src=10.0.10.20 spt=444 deviceInboundInterface=X6-V333 dst=123.45.67.890  dpt=123 deviceOutboundInterface=X1 proto=tcp/12345 in=1500 msg=\"Gateway Anti-Virus Status: SMB file restart detected. File forwarding to Sandbox truncated for filename: hello.xlsx.\" cnt=123456 fw_action=\"NA\" fileid=\"0b9999999999999999ff99\" filetxstatus=230",
+    "event": {
+      "code": "1460",
+      "severity": 5,
+      "category": [
+        "network"
+      ],
+      "kind": "event",
+      "reason": "Gateway Anti-Virus Status: SMB file restart detected. File forwarding to Sandbox truncated for filename: hello.xlsx.",
```
Comment (squioc) [inline SonicWall/sonicwall-fw/tests/test_file_transfer_result.json:20]: I think we can extract the filename from the message with a grok stage and add it to `file.name`

File: SonicWall/sonicwall-fw/tests/test_filename_logging.json [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+    "sekoiaio": {
+      "intake": {
+        "dialect": "SonicWall Firewall",
+        "dialect_uuid": "ee0b3023-524c-40f6-baf5-b69c7b679887"
+      }
+    },
+    "message": "CEF:0|SonicWall|NSa 4500|7.0.1-1234-R5678|1574|Filename Logging|5|cat=0 gcat=3 smac=00:11:22:33:44:55 src=10.0.30.40 spt=12345 deviceInboundInterface=X3-V333 cs3Label=WORKSTATION dmac=66:77:88:99:00:11 dst=22.3.4.55 dpt=444 deviceOutboundInterface=X3-V33 cs4Label=LAN susr=\"USER\" proto=tcp/445 out=32701234 in=31445678 cs5Label=\"Auto\" cs1=\"555 (WORKSTATION->WORK)\" app=9876 msg=\"Filename: FILENAME\" cnt=123456789 fw_action=\"NA\" dpi=1"
+  },
+  "expected": {
+    "message": "CEF:0|SonicWall|NSa 4500|7.0.1-1234-R5678|1574|Filename Logging|5|cat=0 gcat=3 smac=00:11:22:33:44:55 src=10.0.30.40 spt=12345 deviceInboundInterface=X3-V333 cs3Label=WORKSTATION dmac=66:77:88:99:00:11 dst=22.3.4.55 dpt=444 deviceOutboundInterface=X3-V33 cs4Label=LAN susr=\"USER\" proto=tcp/445 out=32701234 in=31445678 cs5Label=\"Auto\" cs1=\"555 (WORKSTATION->WORK)\" app=9876 msg=\"Filename: FILENAME\" cnt=123456789 fw_action=\"NA\" dpi=1",
```
Comment (squioc) [inline SonicWall/sonicwall-fw/tests/test_filename_logging.json:12]: Same as above: we can extract the filename from the message and add it to `file.name`

File: SonicWall/sonicwall-fw/tests/test_syslog_website_accessed.json [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+      "ingress": {
+        "interface": {
+          "name": "X0-V123"
+        }
+      }
+    },
+    "sonicwall": {
+      "fw": {
+        "dpi": true,
+        "fw_action": "drop",
```
Comment (squioc) [inline SonicWall/sonicwall-fw/tests/test_syslog_website_accessed.json:40]: I think we can copy this value and add it to `event.action` and change `event.type` to `['denied']` in this case

**Review by squioc — CHANGES_REQUESTED** [review]
Some changes requested

**Review by squioc — APPROVED** [review]

**Pull request merged**